p2p/src/peers/mod.rs

```diff
@@ -1,85 +1,257 @@
-        let index = thread_rng().gen_range(0, std::cmp::max(self.peers.len(), 1));
-
-        // Get element at index
-        let random_addr = self
-            .peers
-            // get peer infos
-            .values()
-            // enumerate them -> (indices, peer info)
-            .enumerate()
-            // filter by index and get address -> Iterator<Option<SocketAddr>>
-            .filter_map(|(i, v)| if i == index { Some(v.address) } else { None })
-            // Get first one, because
-            .next()
-            .map(|v| v.to_owned());
-
-        Ok(random_addr)
+        let index = thread_rng().gen_range(0, bucket.len());
+
+        bucket.values().nth(index).map(|v| v.address.to_owned())
     }
 
-    /// Get all the peers from the list
-    pub fn get_all(&self) -> Result<Vec<SocketAddr>, failure::Error> {
-        Ok(self.peers.values().map(|v| v.address).collect())
+    /// Get all the peers from the tried bucket
+    pub fn get_all_from_tried(&self) -> Result<Vec<SocketAddr>, failure::Error> {
+        Ok(self.tried_bucket.values().map(|v| v.address).collect())
+    }
+
+    /// Get all the peers from the tried bucket
+    pub fn get_all_from_new(&self) -> Result<Vec<SocketAddr>, failure::Error> {
+        Ok(self.new_bucket.values().map(|v| v.address).collect())
+    }
+
+    pub fn clear_tried_bucket(&mut self) {
+        self.tried_bucket.clear();
+    }
+
+    pub fn clear_new_bucket(&mut self) {
+        self.new_bucket.clear();
+    }
+}
+
+/// Returns the ip and ip split
+fn split_socket_addresses(socket_addr: &SocketAddr) -> (Vec<u8>, Vec<u8>, Vec<u8>) {
+    match socket_addr {
+        SocketAddr::V4(addr) => {
+            let ip = addr.ip().octets();
+            let (left, right) = ip.split_at(ip.len() / 2);
+            (ip.to_vec(), left.to_vec(), right.to_vec())
+        }
+        SocketAddr::V6(addr) => {
+            let ip = addr.ip().octets();
+            let (left, right) = ip.split_at(ip.len() / 2);
+            (ip.to_vec(), left.to_vec(), right.to_vec())
+        }
     }
 }
+
+/// Algorithm to calculate index for the tried addresses buckets
+/// SK = random value chosen when node is born.
+/// IP = the peer’s IP address and port number.
+/// Group = the peer’s group
+/// Host_ID = the peer's host id
+///
+/// i = Hash( SK, IP ) % 4
+/// Bucket = Hash( SK, Group, i ) % 64
+/// Slot = Hash( SK, Host_ID, i ) % 64
+///
+/// Index = Bucket * Slot
+fn calculate_index_for_tried(sk: u8, ip: &[u8], group: &[u8], host_id: &[u8]) -> u16 {
+    let data = [&[sk], &ip[..]].concat();
+    let data_hash = calculate_sha256(&data);
+    let i = data_hash.0[31] % 4;
+
+    let data = [&[sk], group, &[i]].concat();
+    let data_hash = calculate_sha256(&data);
+    let bucket = u16::from(data_hash.0[31]) % 64;
+
+    let data = [&[sk], host_id, &[i]].concat();
+    let data_hash = calculate_sha256(&data);
+    let slot = u16::from(data_hash.0[31]) % 64;
+
+    bucket * slot
```

Comment:
I think this should be something like:

```suggestion
    bucket * NUM_SLOTS + slot
```

Otherwise when the bucket is 0 the index is always 0, the slot is ignored.